File: Samples/Verbum/how-to-embed-a-barcode-into-a-document-e20025-11.1.6-/VB/Extensions/BarCode/BarCodeField.vb
Imports Microsoft.VisualBasic
	Imports System.Collections.Generic
	Imports DevExpress.XtraRichEdit.Fields
	Imports DevExpress.XtraRichEdit.Model
	Imports DevExpress.XtraRichEdit.Utils
Namespace BizPad

	Public Class BarCodeField
		Inherits CalculatedFieldBase
		Public Shared ReadOnly FieldType As String = "BRCODE"
		Protected Overrides ReadOnly Property FieldTypeName() As String
			Get
				Return FieldType
			End Get
		End Property

		Private Shared ReadOnly switchesWithArgument As Dictionary(Of String, Boolean) = CreateSwitchesWithArgument("w", "h", "d", "t", "k", "s", "r", "p", "c", "f", "o", "m", "e")
		Protected Overrides ReadOnly Property SwitchesWithArguments() As Dictionary(Of String, Boolean)
			Get
				Return switchesWithArgument
			End Get
		End Property

		Private barcode As New BarCodeImage()

		Public Overrides Sub Initialize(ByVal switches As InstructionCollection)
			MyBase.Initialize(switches)
			barcode.Initialize(switches)
		End Sub

		Public Overrides Function GetCalculatedValueCore(ByVal sourcePieceTable As PieceTable, ByVal targetModel As DocumentModel, ByVal mailMergeDataMode As MailMergeDataMode, ByVal documentField As Field) As Object

			Dim image As RichEditImage = barcode.CreateImage()
			targetModel.MainPieceTable.InsertInlinePicture(DocumentLogPosition.Zero, image)
			Return targetModel
		End Function
	End Class
End Namespace

File: Samples/Verbum/how-to-embed-a-barcode-into-a-document-e20025-11.1.6-/CS/Extensions/BarCode/BarCodeField.cs
namespace BizPad {
    using System.Collections.Generic;
    using DevExpress.XtraRichEdit.Fields;
    using DevExpress.XtraRichEdit.Model;
    using DevExpress.XtraRichEdit.Utils;

    public class BarCodeField : CalculatedFieldBase {
        public static readonly string FieldType = "BRCODE";
        protected override string FieldTypeName {
            get {
                return FieldType;
            }
        }

        static readonly Dictionary<string, bool> switchesWithArgument = CreateSwitchesWithArgument("w", "h", "d", "t", "k", "s", "r", "p", "c", "f", "o", "m", "e");
        protected override Dictionary<string, bool> SwitchesWithArguments {
            get {
                return switchesWithArgument;
            }
        }

        BarCodeImage barcode = new BarCodeImage();

        public override void Initialize(InstructionCollection switches) {
            base.Initialize(switches);
            barcode.Initialize(switches);
        }

        public override object GetCalculatedValueCore(PieceTable sourcePieceTable,
            DocumentModel targetModel,
            MailMergeDataMode mailMergeDataMode,
            Field documentField) {

            RichEditImage image = barcode.CreateImage();
            targetModel.MainPieceTable.InsertInlinePicture(
                DocumentLogPosition.Zero,
                image);
            return targetModel;
        }
    }
}
